File: test/benchmark/r7rs/python/sumfp.py
#!/usr/bin/env python3
# R7RS Benchmark: sumfp (Python)
# Sum of floats from 0.0 to 100000.0
import time


def run(n):
    i = n
    s = 0.0
    while i >= 0.0:
        s += i
        i -= 1.0
    return s


def main():
    t0 = time.perf_counter_ns()
    result = run(100000.0)
    t1 = time.perf_counter_ns()

    expected = 5000050000.0
    if abs(result - expected) < 1.0:
        print("sumfp: PASS")
    else:
        print(f"sumfp: FAIL result={result}")
    print(f"__TIMING__:{(t1 - t0) / 1e6:.3f}")


main()

File: test/benchmark/r7rs/c2mir/sumfp.c
/* Native C2MIR port of r7rs/sumfp2.ls. */
extern int printf(const char *, ...);

static double run_sumfp(double n) {
    double sum = 0.0;
    while (n >= 0.0) {
        sum += n;
        n -= 1.0;
    }
    return sum;
}

int main(void) {
    double result = run_sumfp(100000.0);
    double diff = result - 5000050000.0;
    if (diff < 0.0) diff = -diff;
    printf(diff < 1.0 ? "sumfp: PASS\n" : "sumfp: FAIL\n");
    return diff >= 1.0;
}
